### DISCUSSION TITLE
Running middleware on matched routes only?

### DISCUSSION THREAD

**Discussion by nickmonad** [tokio-rs/axum#1593] [Q&A] [answered]
Hi there!

Is it possible to run middleware only on routes that match in the router? For example, we have a custom metrics middleware layer, but it creates a lot of noise in the metrics when the path is recorded as a label for public services that get a large amount of spam requests (script kiddies looking for exposed secret files, etc)

Of course, we can make a `Service` based middleware that supports configuration options, where we can "collapse" 404 responses, generating a common path label for all those metrics, keeping our timeseries DB trimmed up a bit. But, that also can create problems down the road, if we want application routes to meaningfully return a 404, say, if a user is not found in the database, where the route is technically valid, but semantically returns "not found".

Hopefully that makes sense! Thanks for the help!

**Comment by davidpdrsn** [comment, marked as answer]
You mean like https://docs.rs/axum/latest/axum/struct.Router.html#method.route_layer?

**Reply by nickmonad** [reply]
Derp... how did I miss that? Thanks for the quick reply.
